linker/Mono.Linker/Utilities.cs

```diff
@@ -0,0 +1,227 @@
+			if (genericInst != null)
+				return MakeGenericType (genericInstanceProvider, genericInst);
+
+			var genericParameter = typeToInflate as GenericParameter;
+			if (genericParameter != null) {
+				if (genericParameter.Owner is MethodReference)
+					return genericParameter;
+
+				var elementType = genericInstanceProvider.ElementType.Resolve ();
+				var parameter = elementType.GenericParameters.Single (p => p == genericParameter);
+				return genericInstanceProvider.GenericArguments [parameter.Position];
+			}
+
+			var functionPointerType = typeToInflate as FunctionPointerType;
+			if (functionPointerType != null) {
+				var result = new FunctionPointerType ();
+				result.ReturnType = InflateGenericType (genericInstanceProvider, functionPointerType.ReturnType);
+
+				for (int i = 0; i < functionPointerType.Parameters.Count; i++)
+				{
+					var inflatedParameterType = InflateGenericType(genericInstanceProvider, functionPointerType.Parameters [i].ParameterType);
+					result.Parameters.Add (new ParameterDefinition (inflatedParameterType));
+				}
+
+				return result;
+			}
+
+			var modifierType = typeToInflate as IModifierType;
+			if (modifierType != null) {
+				var modifier = InflateGenericType (genericInstanceProvider, modifierType.ModifierType);
+				var elementType = InflateGenericType (genericInstanceProvider, modifierType.ElementType);
+
+				if (modifierType is OptionalModifierType) {
+					return new OptionalModifierType (modifier, elementType);
+				}
+
+				return new RequiredModifierType (modifier, elementType);
+			}
+
+			var pinnedType = typeToInflate as PinnedType;
+			if (pinnedType != null) {
+				var elementType = InflateGenericType (genericInstanceProvider, pinnedType.ElementType);
+
+				if (elementType != pinnedType.ElementType)
+					return new PinnedType (elementType);
+
+				return pinnedType;
+			}
+
+			var pointerType = typeToInflate as PointerType;
+			if (pointerType != null) {
+				var elementType = InflateGenericType (genericInstanceProvider, pointerType.ElementType);
+
+				if (elementType != pointerType.ElementType)
+					return new PointerType (elementType);
+
+				return pointerType;
+			}
+
+			var byReferenceType = typeToInflate as ByReferenceType;
+			if (byReferenceType != null) {
+				var elementType = InflateGenericType (genericInstanceProvider, byReferenceType.ElementType);
+
+				if (elementType != byReferenceType.ElementType)
+					return new ByReferenceType (elementType);
+
+				return byReferenceType;
+			}
+
+			var sentinelType = typeToInflate as SentinelType;
+			if (sentinelType != null) {
+				var elementType = InflateGenericType (genericInstanceProvider, sentinelType.ElementType);
+
+				if (elementType != sentinelType.ElementType)
+					return new SentinelType (elementType);
+
+				return sentinelType;
+			}
+
+			return typeToInflate;
+		}
+
+		internal static GenericInstanceType MakeGenericType (GenericInstanceType genericInstanceProvider, GenericInstanceType type)
```

Comment:
If you look at the tests I wrote, the types used are already nested types since they are defined within the top level test case class.

Could you elaborate on what case you think will not be handled?